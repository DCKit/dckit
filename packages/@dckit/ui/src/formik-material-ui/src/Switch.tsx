import * as React from 'react';
import MuiSwitch, {
  SwitchProps as MuiSwitchProps,
} from '@material-ui/core/Switch';
import {
  FieldInputProps,
  FieldMetaProps,
  FieldHelperProps,
  useField,
  useFormikContext,
} from 'formik';

export interface SwitchProps
  extends Omit<
    MuiSwitchProps,
    'checked' | 'name' | 'value' | 'defaultChecked'
  > {
  name: string;
}

export function useFieldToSwitch<Val = unknown>(
  { disabled, name, ...props }: SwitchProps,
  customize?: (
    props: [FieldInputProps<Val>, FieldMetaProps<Val>, FieldHelperProps<Val>]
  ) => Partial<Omit<SwitchProps, 'name'>>
): MuiSwitchProps {
  const { isSubmitting } = useFormikContext();
  const fieldProps = useField(name);

  const [field] = fieldProps;

  return {
    disabled: disabled ?? isSubmitting,
    ...props,
    ...field,
    value: field.name,
    checked: field.value,
    ...customize?.(fieldProps),
  };
}

export function Switch(props: SwitchProps) {
  return <MuiSwitch {...useFieldToSwitch(props)} />;
}

Switch.displayName = 'FormikMaterialUISwitch';
