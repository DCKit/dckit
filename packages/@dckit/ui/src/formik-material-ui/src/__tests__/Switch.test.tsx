import * as React from 'react';
import { Formik, Form } from 'formik';
import renderer from 'react-test-renderer';

import { Switch } from '../Switch';

test('Switch Renders Correctly', () => {
  const component = renderer.create(
    <Formik initialValues={{}} onSubmit={() => {}}>
      <Form>
        <Switch name="test" />
      </Form>
    </Formik>
  );

  expect(component.toJSON()).toMatchSnapshot();
});
