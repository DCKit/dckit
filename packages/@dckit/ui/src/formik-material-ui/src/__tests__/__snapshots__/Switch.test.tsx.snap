// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Renders Correctly 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <span
    className="MuiSwitch-root"
  >
    <span
      aria-disabled={false}
      className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-13 MuiSwitch-switchBase MuiSwitch-colorSecondary"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <span
        className="MuiIconButton-label"
      >
        <input
          className="PrivateSwitchBase-input-16 MuiSwitch-input"
          disabled={false}
          name="test"
          onChange={[Function]}
          type="checkbox"
          value="test"
        />
        <span
          className="MuiSwitch-thumb"
        />
      </span>
    </span>
    <span
      className="MuiSwitch-track"
    />
  </span>
</form>
`;
